### PR TITLE
[tune] Refactor Syncer / deprecate Sync client

### PR DESCRIPTION
**Pull request by krfricke**
## Why are these changes needed?

This PR includes / depends on #25709

The two concepts of Syncer and SyncClient are confusing, as is the current API for passing custom sync functions.

This PR refactors Tune's syncing behavior. The Sync client concept is hard deprecated. Instead, we offer a well defined Syncer API that can be extended to provide own syncing functionality. However, the default will be to use Ray AIRs file transfer utilities.

New API:
- Users can pass `syncer=CustomSyncer` which implements the `Syncer` API
- Otherwise our off-the-shelf syncing is used
- As before, syncing to cloud disables syncing to driver

Changes:
- Sync client is removed
- Syncer interface introduced
- _DefaultSyncer is a wrapper around the URI upload/download API from Ray AIR
- SyncerCallback only uses remote tasks to synchronize data
- Rsync syncing is fully depracated and removed
- Docker and kubernetes-specific syncing is fully deprecated and removed
- Testing is improved to use `file://` URIs instead of mock sync clients

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by krfricke** [comment]
Release tests pass: https://buildkite.com/ray-project/release-tests-pr/builds/6231

**Comment by krfricke** [comment]
Another round of release tests passing: https://buildkite.com/ray-project/release-tests-pr/builds/6322

**Review by Yard1 — COMMENTED** [review]
Do we also need to update AIR RunConfig?

EDIT: looks like we just pass sync config there so it should be good.

File: python/ray/tune/syncer.py [hunk 1]
```diff
@@ -142,395 +79,408 @@ class SyncConfig:
...
+    """Syncer class for synchronizing data between Ray nodes and external storage.
+
+    This class handles data transfer for two cases:
+
+    1. Synchronizing data from the driver to external storage. This affects
+      experiment-level checkpoints and trial-level checkpoints if no cloud storage
+      is used.
+    2. Synchronizing data from remote trainables to external storage.
+
+    Synchronizing tasks are usually asynchronous and can be awaited using ``wait()``.
```
Comment (xwjiang2010) [inline python/ray/tune/syncer.py:144]: May not be in the scope of this PR as I understand async is current behavior.
However, thinking about the whole story of cloud checkpointing, I keep wonder what should be the right semantics to work with (I also tagged you in my PR on the report API, ptal there).

tldr, there will be cases that user saves to a directory in some session, and then want to checkpoint that directory. Who should be responsible for managing the life cycle of that local directory? 
Option 1: our framework. Then we need to make sure that the local directory won't get blown up by deleting stuff regularly. The API is less intuitive, as we are moving away from with Tune.checkpoint_dir(iteration) kind of API, where iteration is used as the index for sub folder, that Tune keeps track of. 
Option 2: User's responsibility. Seems better with the new report API. However, it may be racing if the semantics of syncing to cloud is "asynchronous".

File: python/ray/tune/syncer.py [hunk 2]
```diff
@@ -142,395 +79,408 @@ class SyncConfig:
...
-        self._remote_dir = remote_dir
+class Syncer(abc.ABC):
+    """Syncer class for synchronizing data between Ray nodes and external storage.
+
+    This class handles data transfer for two cases:
+
+    1. Synchronizing data from the driver to external storage. This affects
+      experiment-level checkpoints and trial-level checkpoints if no cloud storage
+      is used.
+    2. Synchronizing data from remote trainables to external storage.
```
Comment (xwjiang2010) [inline python/ray/tune/syncer.py:142]: Slightly off the topic, but do you see if being used in TrainSession as well?

File: python/ray/tune/syncer.py (hunk 1 above)
Comment (xwjiang2010) [reply python/ray/tune/syncer.py:144]: Here is one proposal:

user can always write to "whatever_directory" under the current working directory (which is changed in the beginning)
upon report() returned, our framework makes sure that the move is completed to move "whatever_directory" to "checkpoint_{iter}" directory. The operation should be cheap.
Now, our framework can manage this "checkpoint_{iter}" directory the same way that we do to Tune checkpoints today.
Thus, the restriction on syncing semantics can be removed.
cc @krfricke to cross check with syncer logic.

File: python/ray/tune/syncer.py [hunk 3]
```diff
@@ -142,395 +79,408 @@ class SyncConfig:
...
-
-        Arguments:
-            local_dir: Directory to sync. Uniquely identifies the syncer.
-            remote_dir: Remote directory to sync with.
-            sync_client: Client for syncing between local_dir and
-                remote_dir. Defaults to a Noop.
-        """
-        self._local_dir = os.path.join(local_dir, "") if local_dir else local_dir
-        self._remote_dir = remote_dir
+class Syncer(abc.ABC):
```
Comment (xwjiang2010) [inline python/ray/tune/syncer.py:134]: Overall, I feel like syncing to cloud provides less guarantee than syncing to driver. Is this what we/user want?

File: python/ray/tune/syncer.py [hunk 4]
```diff
@@ -142,395 +79,408 @@ class SyncConfig:
...
 
-    if not remote_dir:
-        _syncers[key] = CloudSyncer(local_dir, remote_dir, NOOP)
-        return _syncers[key]
 
-    if sync_function == "auto":
-        sync_function = None  # Auto-detect
+@DeveloperAPI
+class SyncerCallback(Callback):
+    """Callback to synchronize trial directories on a worker node with the driver."""
```
Comment (xwjiang2010) [inline python/ray/tune/syncer.py:412]: Beef up that this is synchronous?

Comment (xwjiang2010) [reply python/ray/tune/syncer.py:412]: I actually don't quite get why one is sync and one is async

File: python/ray/tune/syncer.py [hunk 5]
```diff
@@ -142,395 +79,408 @@ class SyncConfig:
...
+        except TuneError as e:
+            # Errors occurring during this wait are not fatal for this
+            # checkpoint, so it should just be logged.
+            logger.error(
+                f"Trial {trial}: An error occurred during the "
+                f"checkpoint syncing of the previous checkpoint: {e}"
+            )
+        sync_process.start(
+            source_ip=source_ip,
+            source_path=self._remote_trial_logdir(trial),
```
Comment (xwjiang2010) [inline python/ray/tune/syncer.py:468]: do you know why we are syncing the whole trial logdir instead of this particular checkpoint_00000x folder?

File: python/ray/tune/syncer.py [hunk 6]
```diff
@@ -142,395 +79,408 @@ class SyncConfig:
...
-    def _remove_trial_syncer(self, trial: "Trial"):
-        self._syncers.pop(trial, None)
+        if not force and not self._should_sync(trial):
+            return False
 
-    def _sync_trial_checkpoint(self, trial: "Trial", checkpoint: _TrackedCheckpoint):
-        if checkpoint.storage_mode == CheckpointStorage.MEMORY:
-            return
+        if NODE_IP in trial.last_result:
+            source_ip = trial.last_result[NODE_IP]
```
Comment (xwjiang2010) [inline python/ray/tune/syncer.py:453]: Off the topic as again I know this is current implementation. But I am wondering if this should really come from TrackedCheckpoint? What if we are talking about syncing checkpoint if distributed HuggingFace case? Where the ckpt is at is different than the trial.last_result[NODE_IP].

**Review by xwjiang2010 — APPROVED** [review]
Thanks Kai. This looks great!
Pretty much all my comments are about converging to a future proof checkpointing story. Nothing blocking this PR. But would be beneficial to get your opinion on them!

**Review by sven1977 — APPROVED** [review]
Approved! :)

File: python/ray/tune/syncer.py (hunk 3 above)
Comment (krfricke) [reply python/ray/tune/syncer.py:134]: Why do you feel that way? For syncing to cloud we retry up to three times per default and verbosely log sync failures, so I think we actually do a good job here.

We don't have strong consistency targets right now, but that's true for both syncing mechanisms and something we should work towards to in the future.

File: python/ray/tune/syncer.py (hunk 2 above)
Comment (krfricke) [reply python/ray/tune/syncer.py:142]: Ah yes! This is general purpose enough for syncing with anything

File: python/ray/tune/syncer.py (hunk 1 above)
Comment (krfricke) [reply python/ray/tune/syncer.py:144]: IIUC this will affect how Tune calls the Syncer interface - i.e. which directories it tells the Syncer to sync. From that side any logic would be fine. Let's have the discussion about the actual API in your PR!

File: python/ray/tune/syncer.py (hunk 4 above)
Comment (krfricke) [reply python/ray/tune/syncer.py:412]: There is a `wait` parameter that will tell to wait for syncing to finish (or not).

Actually I guess we could remove the first `wait()` if a sync is in process and `force=False`, let me add this to the PR.

File: python/ray/tune/syncer.py (hunk 6 above)
Comment (krfricke) [reply python/ray/tune/syncer.py:453]: If I understand correctly we will still have a metadata checkpoint at the last_result[NODE_IP] location that then points to the actual location. Also the checkpoint manager does `self.node_ip = node_ip or self.metrics.get(NODE_IP, None)` so I'm wondering if these are actually ever out of sync.

File: python/ray/tune/syncer.py (hunk 5 above)
Comment (xwjiang2010) [reply python/ray/tune/syncer.py:471]: Can you point me to the code?
